tools/base/aio.py

```diff
@@ -83,3 +105,405 @@ async def run_system_command():
+        """
+        return self.submitting or self.running
+
+    @property
+    def closed(self) -> bool:
+        """If an unhandled error occurs, the generator is closed and no further
+        processing should happen
+        """
+        return self.closing_lock.locked()
+
+    @cached_property
+    def closing_lock(self) -> asyncio.Lock:
+        """Flag to indicate whether the generator has been closed"""
+        return asyncio.Lock()
+
+    @async_property
+    async def coros(self) -> AsyncIterator[Union[AsyncParallelIteratorError, Awaitable]]:
+        """An async iterator of the provided coroutines"""
+        coros = self.iter_coros()
+        try:
+            async for coro in coros:
+                yield coro
+        except GeneratorExit:
+            # If we exit before we finish generating we land here (ie error was raised)
+            # In this case we need to tell the (possibly) async generating provider to
+            # also close.
+            try:
+                await coros.aclose()  # type:ignore
+            finally:
+                # Suppress errors closing the provider generator
+                # This can raise a further `GeneratorExit` but it will stop providing.
+                return
+
+    @property
+    def default_limit(self) -> int:
+        """Default is to use cpu+4 to a max of 32 coroutines"""
+        # This reflects the default for asyncio's `ThreadPoolExecutor`, this is a fairly
+        # arbitrary number to use, but it seems like a reasonable default.
+        return min(32, (os.cpu_count() or 0) + 4)
+
+    @cached_property
+    def is_async(self) -> bool:
+        """Provided coros is some kind of async provider"""
+        return isinstance(self._coros, (types.AsyncGeneratorType, AsyncIterator, AsyncIterable))
+
+    @cached_property
+    def is_generator(self) -> bool:
+        """Provided coros is some kind of generator"""
+        return isinstance(self._coros, (types.AsyncGeneratorType, types.GeneratorType))
+
+    @cached_property
+    def limit(self) -> int:
+        """The limit for concurrent coroutines"""
+        return self._limit or self.default_limit
+
+    @cached_property
+    def nolimit(self) -> bool:
+        """Flag indicating no limit to concurrency"""
+        return self.limit == -1
+
+    @cached_property
+    def out(self) -> asyncio.Queue:
+        """Queue of results to yield back"""
+        return asyncio.Queue()
+
+    @property
+    def running(self) -> bool:
+        """Flag to indicate whether any tasks are running"""
+        return not self.running_queue.empty()
+
+    @cached_property
+    def running_queue(self) -> asyncio.Queue:
+        """Queue which is incremented/decremented as tasks begin/end
+
+        This is for tracking when there are no longer any tasks running.
+
+        A queue is used here as opposed to other synchronization primitives, as
+        it allows us to get the size and emptiness.
```

Comment:
Again, it's a bit unclear to me what is internal to this interface and what isn't. I would really like to see a clean interface with the basic end user facing primitives and understand what is internal. I realize Python encourages the notion that everyone is a responsible adult who should be able to peak inside, but having some way to know what is really an internal implementatoin detail while reading this code would be good.

Reply:
its intended use is functionally - ie 

```
async for result in parallel(awaitable_generator):
    await do_something_with(result)
```

so in that sense, everything is internal to it, in much the same way as that would be the case with for example a decorator

Reply:
python doesnt really have a strong distinction of public/private variables

sure there is the convention of using `_` to mark something as "private", but in the python world that semantically means - "if you subclass this, and make use of this private var, your implementation might break in the future"

(i understand why you dislike my habit of using `_` for local vars in this respect, its just a habit of years, and it provides convenience)

in that sense, i guess these methods are public to subclass implementations, but i dont forsee any implemented usage that would access these vars directly

Reply:
I think given we have a parallel ABC discussion, we can land this PR as is.